velox/functions/prestosql/tests/ArrayNGramTest.cpp

```diff
@@ -0,0 +1,99 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <optional>
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+using namespace facebook::velox::test;
+using namespace facebook::velox::functions::test;
+
+namespace {
+
+class ArrayNGramTest : public FunctionBaseTest {
+ protected:
+  template <typename T>
+  void testNgram(
+      const std::vector<std::vector<std::optional<T>>>& inputArray,
+      int64_t size,
+      const std::vector<std::optional<std::vector<std::optional<T>>>>&
+          expectedOutput) {
+    auto input = makeNullableArrayVector<T>({inputArray});
+    auto result =
+        evaluate(fmt::format("ngrams(c0, {})", size), makeRowVector({input}));
+
+    auto expected = makeNullableNestedArrayVector<T>({{expectedOutput}});
+    assertEqualVectors(expected, result);
+  }
+};
+
+TEST_F(ArrayNGramTest, intArrays) {
+  testNgram<int64_t>({{1, 2, 3, 4}}, 1, {{{1}}, {{2}}, {{3}}, {{4}}});
+  testNgram<int64_t>({{1, 2, 3, 4}}, 2, {{{1, 2}}, {{2, 3}}, {{3, 4}}});
+  testNgram<int64_t>({{1, 2, 3, 4}}, 3, {{{1, 2, 3}}, {{2, 3, 4}}});
+  testNgram<int64_t>({{1, 2, 3, 4}}, 4, {{{1, 2, 3, 4}}});
+  testNgram<int64_t>({{1, 2, 3, 4}}, 5, {{{1, 2, 3, 4}}});
+  testNgram<int64_t>(
+      {{1, 2, 3, 4}}, std::numeric_limits<int32_t>::max(), {{{1, 2, 3, 4}}});
+  testNgram<int64_t>(
+      {{1, 2, 3, 4}},
+      std::numeric_limits<int32_t>::max() + (int64_t)(1000),
+      {{{1, 2, 3, 4}}});
+  testNgram<int64_t>({{}}, 1, {{{}}});
+  testNgram<int64_t>({{}}, 10, {{{}}});
+
+  // Test of invalid N
+  auto input = makeArrayVector<int64_t>({{1, 2, 3, 4}});
+  VELOX_ASSERT_THROW(
+      evaluate("ngrams(c0, 0)", makeRowVector({input})), "N must be positive.");
+  VELOX_ASSERT_THROW(
+      evaluate("ngrams(c0, -5)", makeRowVector({input})),
+      "N must be positive.");
+  input = makeArrayVector<int64_t>({{}});
+  VELOX_ASSERT_THROW(
+      evaluate("ngrams(c0, 0)", makeRowVector({input})), "N must be positive.");
+}
+
+TEST_F(ArrayNGramTest, stringArrays) {
+  testNgram<std::string>(
+      {{"foo", "bar", "baz", "foo"}},
+      1,
+      {{{"foo"}}, {{"bar"}}, {{"baz"}}, {{"foo"}}});
+  testNgram<std::string>(
+      {{"foo", "bar", "baz", "foo"}},
+      2,
+      {{{"foo", "bar"}}, {{"bar", "baz"}}, {{"baz", "foo"}}});
+  testNgram<std::string>(
+      {{"foo", "bar", "baz", "foo"}},
+      3,
+      {{{"foo", "bar", "baz"}}, {{"bar", "baz", "foo"}}});
+  testNgram<std::string>(
+      {{"foo", "bar", "baz", "foo"}}, 4, {{{"foo", "bar", "baz", "foo"}}});
+  testNgram<std::string>(
+      {{"foo", "bar", "baz", "foo"}}, 5, {{{"foo", "bar", "baz", "foo"}}});
+}
+
+TEST_F(ArrayNGramTest, testNull) {
+  testNgram<std::string>(
```

Comment:
Please add tests where n is null, input array is null and both are null as well.

Reply:
This is not necessary as we are using `void call` and default NULL behavior. In this case, if any input is NULL, output is already NULL. More details here https://facebookincubator.github.io/velox/develop/scalar-functions.html#null-behavior

Most of the functions have such behavior and skip NULL argument tests.